hilti/include/rt/types/stream.h

```diff
@@ -112,65 +142,230 @@ class Chunk {
-        while ( i && i->_next )
-            i = i->_next;
-        return i;
+    // Link in chunk as successor of current one. Updates offset/chain for
+    // the appended chunk and all its successors.
+    void setNext(std::unique_ptr<Chunk> next) {
+        assert(_chain);
+
+        Offset offset = endOffset();
+        _next = std::move(next);
+
+        auto c = _next.get();
+        while ( c ) {
+            c->_offset = offset;
+            c->_chain = _chain;
+            offset += c->size();
+            c = c->_next.get();
+        }
     }
 
     void clearNext() { _next = nullptr; }
-    void setNext(std::shared_ptr<Chunk> c) { _next = std::move(c); }
-    void setOffset(Offset o) { _offset = o; }
-    bool tryAppend(const Chunk& d); // Appends to small buffer is possible, returns false of not.
-    void trim(Offset o);
-
-    void debugPrint(std::ostream& out) const;
 
 private:
-    // Note: We must not have a pointer to the parent stream instance in
-    // chunks because the parent may be on the stack with a shorter life
-    // time.
-    Offset _offset = 0;
-    std::variant<Array, Vector> _data;
-    std::shared_ptr<Chunk> _next = nullptr;
-    bool _frozen = false;
+    Chunk() : _data(Array()) {}
 
-    // TODO(robin): Implement later.
-    //  std::optional<Object> object;
-    //  std::vector<std::int64_t> marks; // offsets relative to this chunk
+    template<int N>
+    Chunk _fromArray(Offset o, std::array<Byte, N> d) {
+        if constexpr ( N <= Chunk::SmallBufferSize )
+            return Chunk(o, std::move(d));
+
+        return Chunk(o, Chunk::Vector(d.begin(), d.end()));
+    }
+
+
+    inline Chunk _fromArray(Offset o, const char* d, Size n) {
+        auto ud = reinterpret_cast<const Byte*>(d);
+
+        if ( n <= Chunk::SmallBufferSize ) {
+            std::array<Byte, Chunk::SmallBufferSize> x{};
+            std::copy(ud, ud + n.Ref(), x.data());
+            return Chunk(o, std::move(x), n);
+        }
+
+        return Chunk(o, Chunk::Vector(ud, ud + n.Ref()));
+    }
+
+    Offset _offset = 0;                // global offset of 1st byte
+    std::variant<Array, Vector> _data; // content of this chunk
+    const Chain* _chain = nullptr;     // chain this chunk is part of, or null if not linked to a chain yet (non-owning;
+                                       // will stay valid at least as long as the current chunk does)
+    std::unique_ptr<Chunk> _next = nullptr; // next chunk in chain, or null if last
 };
 
-/** The main content structure for a heap-allocated stream object. */
-struct Chain {
-    std::shared_ptr<Chunk> head;
-    std::shared_ptr<Chunk> tail;
+/**
+ * Main data structure for the content of a stream object. A chain is
+ * heap-allocated by the stream and retains ownership of the linked chunks. A
+ * Chain may survive its stream in the case that iterators to any of its
+ * chunks are still around.
+ */
+class Chain : public intrusive_ptr::ManagedObject {
```

Comment:
This is effectively a list with extra offset and state information now, and it should be possible to delegate all the list behavior to some `std::list` or `std::deque` member. That would get rid of a lot of low-level twiddling, but involve some more work disentangling what is happening here.

Reply:
Prefer to leave as it as that gives us more control.